Public/Src/Cache/ContentStore/Distributed/Stores/DistributedContentCopier.cs

```diff
@@ -359,35 +360,50 @@ void traceCopyFailed(Context c)
-                                _contentLocationStore.MachineReputationTracker.ReportReputation(location, MachineReputation.Bad);
+                                _contentLocationStore.ReportReputation(location, MachineReputation.Bad);
                                 break;
                             default:
-                                return (new ErrorResult(copyFileResult, $"{AttemptTracePrefix(attemptCount)} File copier result code {copyFileResult.Code} is not recognized").AsResult<PutResult>(), true);
+                                return (result: new ErrorResult(copyFileResult, $"{AttemptTracePrefix(attemptCount)} File copier result code {copyFileResult.Code} is not recognized").AsResult<PutResult>(), retry: true);
                         }
 
                         if (copyFileResult.Succeeded)
                         {
+                            // The copy succeeded, but it is possible that the resulting size doesn't match an expected one.
+                            if (hashInfo.Size != -1 && copyFileResult.Size != null && hashInfo.Size != copyFileResult.Size.Value)
+                            {
+                                lastErrorMessage =
+                                    $"Contenthash {hashInfo.ContentHash} at location {location} has content size {copyFileResult.Size.Value} mismatch from {hashInfo.Size}";
+                                Tracer.Warning(
+                                    context,
+                                    $"{AttemptTracePrefix(attemptCount)} {lastErrorMessage} Trying another replica.");
+                                badContentLocations.Add(location);
+                                continue;
+                            }
+
                             PutResult putResult = await handleCopyAsync((copyFileResult, tempLocation, attemptCount));
 
                             if (putResult.Succeeded)
                             {
                                 // The put succeeded, but this doesn't necessarily mean that we put the content we intended. Check the content hash
                                 // to ensure it's what is expected. This should only go wrong for a small portion of non-trusted puts.
-                                if (putResult.ContentHash == hashInfo.ContentHash)
-                                {
-                                    // Don't delete the temporary file! It no longer exists after the Put moved it into the cache
-                                    deleteTempFile = false;
-                                    return (putResult, false);
-                                }
-                                else
+                                if (putResult.ContentHash != hashInfo.ContentHash)
                                 {
+                                    lastErrorMessage =

```

Comment:
>lastErrorMessage [](start = 36, length = 16)

Since you are looping over a set of locations  would it be helfpull if the location that has an issue was part of the message? Also perhaps this should be a telemetry counter so we can keep an eye out?

Reply:
When the issue occurs we trace the location there. Let me think about whether it's useful to have the location in the final result. (I think you're right).